### ISSUE TITLE
unused variable false positive in lazy nested scope

### ISSUE THREAD

**Issue by ekorman** [astral-sh/ty#3442]
### Summary

for this code block
```python
def foo1(i: int):

    def _bar():
        return [[k for k in range(i)] for _ in range(2)]

    return _bar


def foo2(i: int):

    def _bar():
        return [k for k in range(i)]

    return _bar


def foo3(i: int):
    return [[k for k in range(i)] for _ in range(2)]
```

the extension says incorrectly that parameter `i` is unused in `foo1`, while `foo2` and `foo3` do not have this problem

### Version

0.0.35

**Comment by MichaReiser** [comment]
We fail to mark the binding as used because `bar` in `foo1` is lazy. It seems somewhat problematic that our used marking is coupled to our snapshotting mechanism.

**Comment by carljm** [comment]
Simpler repro (from #3590):

```py
def outer() -> None:
    a = 1

    def inner() -> list[int]:
        return [a + x for x in [1, 2, 3]]
        # return [a + 1]  # not reported unused if used outside of a comprehension
```

**Comment by denyszhak** [comment]
@MichaReiser For this fix, I think it makes sense to patch over the existing approach and we could have a separate issue for your concern to consider decoupling the two?

**Comment by denyszhak** [comment]
@MichaReiser As part of the fix, I tried decoupling usage tracking from snapshot contents in the linked PR. It still placed into `update_lazy_snapshots` for later reassignments, but it no longer reads bindings out of the snapshot payload, it marks the current live bindings directly.
